riscv/csrs.cc

```diff
@@ -211,22 +240,84 @@ bool pmpcfg_csr_t::unlogged_write(const reg_t val) noexcept {
     return false;
 
   bool write_success = false;
+  bool rlb = state->mseccfg->get_rlb();
+  bool mml = state->mseccfg->get_mml();
   for (size_t i0 = (address - CSR_PMPCFG0) * 4, i = i0; i < i0 + proc->get_xlen() / 8; i++) {
     if (i < proc->n_pmp) {
-      if (!(state->pmpaddr[i]->cfg & PMP_L)) {
+      bool locked = (state->pmpaddr[i]->cfg & PMP_L);
+      if (rlb || (!locked && !state->pmpaddr[i]->next_locked_and_tor())) {
         uint8_t cfg = (val >> (8 * (i - i0))) & (PMP_R | PMP_W | PMP_X | PMP_A | PMP_L);
-        cfg &= ~PMP_W | ((cfg & PMP_R) ? PMP_W : 0); // Disallow R=0 W=1
+        // Drop R=0 W=1 when MML = 0
+        // Remove the restriction when MML = 1
+        if (!mml) {
+          cfg &= ~PMP_W | ((cfg & PMP_R) ? PMP_W : 0); 
+        }
+        // Disallow A=NA4 when granularity > 4
         if (proc->lg_pmp_granularity != PMP_SHIFT && (cfg & PMP_A) == PMP_NA4)
-          cfg |= PMP_NAPOT; // Disallow A=NA4 when granularity > 4
-        state->pmpaddr[i]->cfg = cfg;
+          cfg |= PMP_NAPOT; 
+        /*
+        Adding a rule with executable privileges that either is M-mode-only or a locked Shared-Region is not possible 
+        and such pmpcfg writes are ignored, leaving pmpcfg unchanged. 
+        This restriction can be temporarily lifted e.g. during the boot process, by setting mseccfg.RLB.
+        */ 
+        if (rlb || !(mml && ((cfg & PMP_L) && ((cfg & PMP_X) || ((cfg & PMP_W) && !(cfg & PMP_R)))))) 
+          state->pmpaddr[i]->cfg = cfg;
       }
+
       write_success = true;
     }
   }
   proc->get_mmu()->flush_tlb();
   return write_success;
 }
 
+// implement class mseccfg_csr_t
+mseccfg_csr_t::mseccfg_csr_t(processor_t* const proc, const reg_t addr):
+  csr_t(proc, addr),
+  mseccfg_val(0) {
+}
+
+bool mseccfg_csr_t::get_mml() const noexcept {
+  return (mseccfg_val & MSECCFG_MML); 
+}
+
+bool mseccfg_csr_t::get_mmwp() const noexcept {
+  return (mseccfg_val & MSECCFG_MMWP);
+}
+
+bool mseccfg_csr_t::get_rlb() const noexcept {
+  return (mseccfg_val & MSECCFG_RLB);
+}
+
+reg_t mseccfg_csr_t::read() const noexcept {
+  return mseccfg_val;
+}
+
+bool mseccfg_csr_t::unlogged_write(const reg_t val) noexcept {
+  if (proc->n_pmp == 0)
+    return false;
+
+  bool pmplock_recorded = false;
+  for (size_t i = 0; i < proc->n_pmp; i++) {
+    if (state->pmpaddr[i]->is_locked()) {
+      pmplock_recorded = true;
+      break;
+    }
+  }
```

Comment:
Look into `std::any_of` for a much simpler way to do this.

Reply:
how about following? I would like to keep the is_locked() function since I don't hope to use PMP_L bit directly here.
  bool pmplock_recorded = std::any_of(state->pmpaddr, state->pmpaddr + proc->n_pmp,
          [](pmpaddr_csr_t_p & c) { return c->is_locked(); } );